src/linux/mod.rs

```diff
@@ -25,17 +26,42 @@ pub struct LinuxSandbox {
     full_env: bool,
 }
 
-impl Sandbox for LinuxSandbox {
-    fn new() -> Result<Self> {
-        // Setup landlock filtering.
-        let mut landlock = Ruleset::new()
-            .set_best_effort(false)
-            .handle_access(AccessFs::from_all(ABI))?
-            .create()?;
-        landlock.as_mut().set_no_new_privs(true);
+impl LinuxSandbox {
+    /// Create a customized Linux sandbox.
+    ///
+    /// The [`min_landlock_abi`] argument defines the minimum Landlock Kernel
+    /// ABI version which must be supported. Sandboxing will fail on systems
+    /// which do not support this.
+    ///
+    /// All landlock ABI versions after [`min_landlock_abi`] versions are used
+    /// on systems that support them, but are ignored otherwise. This means
+    /// that the sandbox will be created without any error even if these are
+    /// not supported.
+    pub fn new_with_version(min_landlock_abi: LANDLOCK_ABI) -> Result<Self> {
+        let mut ruleset = Ruleset::new();
+
+        // Require at least `min_landlock_abi`.
+        (&mut ruleset).set_compatibility(CompatLevel::HardRequirement);
+        (&mut ruleset).handle_access(AccessFs::from_all(min_landlock_abi))?;
+
+        // Add optional checks for everything after `min_landlock_abi`.
+        //
+        // NOTE: This will require these access permissions on systems that support
+        // checking for them, while ignoring them on all other systems.
+        (&mut ruleset).set_compatibility(CompatLevel::BestEffort);
+        (&mut ruleset).handle_access(BitFlags::<AccessFs>::all())?;
```

Comment:
How will this work with `add_exception()`? Particularly when it adds rules using `AccessFs::from_write(ABI)` (which will never include `AccessFs::Refer`). Isn't that a problem?

I think a max supported landlock ABI is needed to ensure that `add_exception()` is re-enabling the appropriate permissions. Otherwise, a new landlock ABI version could cause more restriction here that we don't account for in `add_exception()` (e.g., truncate in Landlock v3).

Reply:
> Particularly when it adds rules using AccessFs::from_write(ABI) (which will never include AccessFs::Refer). Isn't that a problem?

It is, good catch. I'm not actually sure what happens when this specifies a newer version than supported by the sandbox. Hopefully our CI tests should be able to verify this works when just using V3.

Reply:
CI passes with V3, so I'd assume using the latest versions in `add_exception` is fine.

Reply:
This is exactly why the [doc says](https://landlock.io/rust-landlock/landlock/enum.AccessFs.html#warning): don’t use `BitFlags::<AccessFs>::all()` nor `BitFlags::ALL` :wink: 

I'd like to wrap these `BitFlags` to make this kind of call impossible at build time, but we're not there yet. See landlock-lsm/rust-landlock#23 too

You should test with a specific Landlock ABI (the latest), test with a kernel that support this version (you can make sure the CI does support all the requested access right by setting early `ruleset.set_compatibility(CompatLevel:Strict)`), and stick to it. If some access rights are required (e.g. `Refer`), you can also specify that with `set_compatibility()`. I'm working on making it simpler, the foundation is merged but the interface is not ready yet: https://github.com/landlock-lsm/rust-landlock/pull/12#issuecomment-1417834004

Reply:
The docs here could probably be clarified, considering I don't really understand what is meant by "compile time behavior at run time, which may look like undefined behavior".

But the usage of "all" here over a specific version is intentional as a way to provide forward-compatibility. Doing so would allow Birdcage to support a new filesystem capability in a hypothetical landlock v4 without having to explicitly update the Birdcage crate every time.

It's difficult to say what will come in the future of course, but looking at previous landlock versions it seems like always using the latest version supported by the kernel is most likely to produce the desired results, even if Birdcage isn't aware of it yet.

> You should test with a specific Landlock ABI (the latest), test with a kernel that support this version (you can make sure the CI does support all the requested access right by setting early ruleset.set_compatibility(CompatLevel:Strict)), and stick to it. If some access rights are required (e.g. Refer), you can also specify that with set_compatibility(). I'm working on making it simpler, the foundation is merged but the interface is not ready yet: https://github.com/landlock-lsm/rust-landlock/pull/12#issuecomment-1417834004

Considering that our consumer for Birdcage (a single application) requires at least 2 different compatibility modes (with and without refer), and needs to work both on V1 and V2+, makes it non-trivial to setup a CI and test harness to verify all these combinations. This is definitely something I want to get done, rather than doing so manually, but it will probably take some time.

Reply:
> Doing so would allow Birdcage to support a new filesystem capability in a hypothetical landlock v4 without having to explicitly update the Birdcage crate every time.

But this is not real "support"... If landlock v4 allows blocking some additional system call, but `add_exception()` still uses `AccessFs::from_write(LANDLOCK_ABI::V3)` to make the exception, that means we will start blocking a new system call with no way to allow it again ...

Not only is that a breaking change. Blocking a system call with no way to add an exception is what created our current issues with Landlock v1 and FS_REFER... And our only solution to that is to disable Landlock.

That does not seem like a good position to put our future selves in.

If we use `AccessFs::from_all(LANDLOCK_ABI::V3)` here, we can support Landlock v4 with a quick update and manual confirmation that our exceptions still make sense.

Reply:
> But this is not real "support"... If landlock v4 allows blocking some additional system call, but add_exception() still uses AccessFs::from_write(LANDLOCK_ABI::V3) to make the exception, that means we will start blocking a new system call with no way to allow it again ...

Yes, but that's only because `from_write` only supports specifying a fixed ABI version. 

> Not only is that a breaking change. Blocking a system call with no way to add an exception is what created our current issues with Landlock v1 and FS_REFER... And our only solution to that is to disable Landlock.

If a new landlock option was introduced that is blocked by default, that would be the only option anyway.

> If we use AccessFs::from_all(LANDLOCK_ABI::V3) here, Birdcage will simply continue to allow that syscall. And we can add real support for Landlock v4 with a quick update and manual confirmation that our exceptions still make sense.

Not with things like FS_REFER.

Reply:
> > The docs here could probably be clarified, considering I don't really understand what is meant by "compile time behavior at run time, which may look like undefined behavior".

Right, it's not clear in this summary, but the refered [ABI documentation](https://landlock.io/rust-landlock/landlock/enum.ABI.html) is better. I created a PR to improve this: landlock-lsm/rust-landlock#50

> > 
> > But the usage of "all" here over a specific version is intentional as a way to provide forward-compatibility. Doing so would allow Birdcage to support a new filesystem capability in a hypothetical landlock v4 without having to explicitly update the Birdcage crate every time.
> > 
> > It's difficult to say what will come in the future of course, but looking at previous landlock versions it seems like always using the latest version supported by the kernel is most likely to produce the desired results, even if Birdcage isn't aware of it yet.

Because we don't know what will come in the future, this is not forward-compatibility but hope-we-get-new-restrictions-but-without-any-guarantee. :wink: 
Forward-compatibility is only guaranteed if we stick to a specific Landlock ABI.

> > But this is not real "support"... If landlock v4 allows blocking some additional system call, but add_exception() still uses AccessFs::from_write(LANDLOCK_ABI::V3) to make the exception, that means we will start blocking a new system call with no way to allow it again ...
> 
> Yes, but that's only because `from_write` only supports specifying a fixed ABI version.

Correct, and this is on purpose to avoid undefined (because not implemented yet) behaviors.

> 
> > Not only is that a breaking change. Blocking a system call with no way to add an exception is what created our current issues with Landlock v1 and FS_REFER... And our only solution to that is to disable Landlock.
> 
> If a new landlock option was introduced that is blocked by default, that would be the only option anyway.

Landlock will always restrict the same way, whatever kernel version is used (older or newer), for the same set of syscalls and arguments (taking into account any call used to build/describe the restriction). This is part of the contract the kernel has with user space. This is really important for the kernel interfaces to be backward compatible, and also forward compatible. So, no new Landlock option will ever block anything by default, we'll always need to opt-in for this new option to take effect.

If you choose to ask for an unknown and untested set of restrictions, there is good chance that you'll get into troubles. I guess you'll prefer to test that a specific new option works as expected (e.g. thanks to the CI) and then guarantee to users that it is safe to update Birdcage or the kernel because they will only get more (vetted and then working) sandboxing.

> 
> > If we use AccessFs::from_all(LANDLOCK_ABI::V3) here, Birdcage will simply continue to allow that syscall. And we can add real support for Landlock v4 with a quick update and manual confirmation that our exceptions still make sense.
> 
> Not with things like FS_REFER.

FS_REFER is not enabled by default, neither by the kernel nor rust-landlock. This can only happen if the handled accesses include FS_REFER, for instance using `BitFlags::<AccessFs>::all()`, which is undefined and should then not be used. This is why a new Landlock ABI must be tested with a kernel supporting this ABI before being used, hence landlock-lsm/rust-landlock#41.